arborist/engine.go

```diff
@@ -0,0 +1,651 @@
+func (engine *Engine) readRoleFromJSON(roleJSON *RoleJSON) (*Role, error) {
+	permissions := make(map[*Permission]struct{})
+	for _, permissionJSON := range roleJSON.Permissions {
+		permission := engine.readPermissionFromJSON(permissionJSON)
+		permissions[permission] = struct{}{}
+	}
+	role := Role{
+		id:          roleJSON.ID,
+		description: roleJSON.Description,
+		permissions: permissions,
+	}
+	return &role, nil
+}
+
+func (engine *Engine) addRole(role *Role) (*Role, error) {
+	if _, exists := engine.roles[role.id]; exists {
+		err := alreadyExists("role", "id", role.id)
+		return nil, err
+	}
+	engine.roles[role.id] = role
+
+	return role, nil
+}
+
+func (engine *Engine) addRoleFromJSON(roleJSON *RoleJSON) (*Role, error) {
+	role, err := engine.readRoleFromJSON(roleJSON)
+	if err != nil {
+		return nil, err
+	}
+	return engine.addRole(role)
+}
+
+// updateRoleWithJSON finds a role with the given ID in the engine and uses the
+// contents of the RoleJSON to overwrite the fields in the existing role with
+// the new stuff from the JSON.
+func (engine *Engine) updateRoleWithJSON(roleID string, roleJSON *RoleJSON) (*Role, error) {
+	role, exists := engine.roles[roleID]
+	if !exists {
+		err := notExist("role", "id", roleID)
+		return nil, err
+	}
+
+	roleJSON.defaultsFromRole(role)
+
+	updatedRole, err := engine.readRoleFromJSON(roleJSON)
+	if err != nil {
+		return nil, err
+	}
+
+	*role = *updatedRole
+
+	return role, nil
+}
+
+func (engine *Engine) appendRoleWithJSON(roleID string, roleJSON *RoleJSON) (*Role, error) {
+	role, exists := engine.roles[roleID]
+	if !exists {
+		err := notExist("role", "id", roleID)
+		return nil, err
+	}
+
+	updatedRole, err := engine.readRoleFromJSON(roleJSON)
+	if err != nil {
+		return nil, err
+	}
+	role.appendFrom(updatedRole)
+
+	return role, nil
+}
+
+// checkDeleteRole checks if it's fine to delete this role from the engine. It
+// is NOT fine if there are policies which refer to this role, in which case it
+// returns an error.
+func (engine *Engine) checkDeleteRole(role *Role) error {
+	for _, policy := range engine.policies {
+		for policyRole := range policy.roles {
+			if role.id == policyRole.id {
+				return noDelete(
+					"role",
+					"id",
+					role.id,
+					fmt.Sprintf("in use by policy %s", policy.id),
+				)
+			}
+		}
+	}
+	return nil
+}
+
+// removeRole deletes the given role from the engine. An error is returned if
+// deleting the role would leave the engine in an incorrect state, namely if
+// there are any policies referencing this role. Policies using the role should
+// be deleted before the role is.
+func (engine *Engine) removeRole(role *Role) error {
+	if role == nil {
+		return nil
+	}
+	err := engine.checkDeleteRole(role)
+	if err != nil {
+		return err
+	}
+	delete(engine.roles, role.id)
+	return nil
+}
+
+// Operations for working with resources
+
+func (engine *Engine) listResourcePaths() []string {
+	result := make([]string, 0)
```

Comment:
This might not be a big deal, but could be a good optimization for future when you have very large data. You can describe the capacity of the slice when you declare it. By default this `make()` will give you a slice of capacity zero, and effectively double it every time you go beyond the capacity using an append. That also means that a new area of contiguous memory is allocated, and your old memory is freed, every time you break the cap.

It's an easy optimization here because you know resulting size of the slice.

Here's some code to demonstrate what I mean:
https://play.golang.org/p/0kTonL2p7MH

Or you can just declare the length of a slice and fill it in using an iterator:
https://play.golang.org/p/rDvnPVU3Csd

Reply:
Yeah, good point! I realized at some point that I hadn't been doing this consistently...python habits I guess. :)

Reply:
Fixed in 221b62f